pycqed/measurement/waveform_control/pulse_library.py

```diff
@@ -113,6 +113,188 @@ def hashables(self, tstart, channel):
+                    'trans_amplitude2', 'trans_length', 'buffer_length_start',
+                    'buffer_length_end', 'channel_relative_delay']:
+            self.add_value_change_callback(par, self._update_lengths_amps)
+        for par in ['channel', 'channel2']:
+            self.add_value_change_callback(par, self._update_channels)
+
+        self._update_lengths_amps()
+        self._update_channels()
+
+    @classmethod
+    def pulse_params(cls):
+        """
+        Returns a dictionary of pulse parameters and initial values.
+        These parameters are set upon calling the
+        super().__init__ method.
+        """
+        params = {
+            'pulse_type': 'NZTransitionControlledPulse',
+            'channel': None,
+            'channel2': None,
+            'amplitude': 0,
+            'amplitude2': 0,
+            'amplitude_offset': 0,
+            'amplitude_offset2': 0,
+            'pulse_length': 0,
+            'trans_amplitude': 0,
+            'trans_amplitude2': 0,
+            'trans_length': 0,
+            'buffer_length_start': 30e-9,
+            'buffer_length_end': 30e-9,
+            'channel_relative_delay': 0,
+            'gaussian_filter_sigma': 1e-9,
+        }
+        return params
+
+    def _update_channels(self):
+        self.channels = [self.channel, self.channel2]
+
+    def _update_lengths_amps(self):
+        self.lengths = []
+        self.amplitudes = []
+
+        for ma, ta, ao, d in [
+            (self.amplitude, self.trans_amplitude, self.amplitude_offset,
+             -self.channel_relative_delay/2),
+            (self.amplitude2, self.trans_amplitude2, self.amplitude_offset2,
+             self.channel_relative_delay/2),
+        ]:
+            ml = self.pulse_length
+            tl = self.trans_length
+            bs = self.buffer_length_start
+            be = self.buffer_length_end
+            self.amplitudes.append([0, ma + ao, ta, -ta, -ma + ao, 0])
+
+            if ta == 0:
+                self.lengths.append([bs + d, ml / 2, tl / 2,
+                                     tl / 2, ml / 2, be - d])
+            else:
+                if np.abs(tl * ta) < np.abs(ml * ao):
+                    raise ValueError('NZTCPulse: Pick the pulse parameters '
+                                     'such that "trans_len * trans_amplitude < '
+                                     'pulse_length * amplitude_offset".')
+                self.lengths.append([bs + d, ml / 2, (tl - ml * ao / ta) / 2,
+                                     (tl + ml * ao / ta) / 2, ml / 2, be - d])
+
+    def add_value_change_callback(self, param, callback):
+        def getter(self):
+            return getattr(self, '_' + param)
+
+        def setter(self, value):
+            setattr(self, '_' + param, value)
+            callback()
+
+        setattr(self, '_' + param, getattr(self, param))
+        setattr(self.__class__, param, property(getter, setter))
```

Comment:
This is called in the init of each instance of the class, but it modifies the class and not only instance. Is this really desirable? (It should not do harm in this case since the getter/setter functions of all instances do the same thing, but it still seems strange to me.)

Reply:
It is not desirable, but I'm not sure if I can set the property for the instance only. I will check